File: Hyungjoo/2805.py
n, m = map(int, input().split())
trees = list(map(int, input().split()))

ans = 0


def solution(cur_height, max_height, min_height):
    global n, m
    global trees
    global ans

    if cur_height == max_height or cur_height == min_height:
        return

    # calculate total acquired tree weights
    tree_value = 0
    for tree in trees:
        if tree > cur_height:
            tree_value += tree - cur_height

    # if total value higher than m and current height higher than previous cache
    if tree_value >= m and cur_height > ans:
        ans = cur_height

    # binary search
    if tree_value >= m:
        solution((cur_height + max_height) // 2, max_height, cur_height)
    else:
        solution((cur_height + min_height) // 2, cur_height, min_height)


solution(max(trees) // 2, max(trees), 0)
print(ans)

File: Ikjae/2805.cpp
#include <bits/stdc++.h>
using namespace std;
long long n, m, tree[1'000'000];
int main(){
	cin >> n >> m;
	long long i, l, r, mid, total, ans;
	l = r = ans = 0;
	for(i=0;i<n;++i) cin >> tree[i];
	for(i=0;i<n;++i) r = max(r, tree[i]);
	while(l <= r){
		mid = (l + r) / 2;
		total = 0;
		for(i=0;i<n;++i){
			if(tree[i] > mid) total += tree[i] - mid;
		}
		if(total >= m){
			ans = max(ans, mid);
			l = mid + 1;
		}else{
			r = mid - 1;
		}
	}
	cout << ans;
	return 0;
}
